### PR TITLE
Data pipeline

### PR DESCRIPTION
**Pull request by j-nystrom**
This is a pretty big PR, too many different things for best practice, so apologies in advance :) Reason is that it's a mix of general repo setup stuff and (almost) all the code to this point for the data processing. 

I'll do my best to guide the review below, starting with what you can skip ;)
- You can pretty much ignore the files related to the overall setup of the repository, pre-commit hooks, formatting and linting (`pre-commit-config.yaml`, `setup.cfg`, etc)
- `general_utils.py` is just code for setting up the `logger` right now, so not very interesting either
- You already saw the DAG setup in the `dags` directory, so you can skip that too
- FInally, you can ignore the `notebook` updates

The interesting stuff is in the `data` folder:
- The `predicts_task.py` is straightforward, so it doesn't require a detailed review
- Where I´d like to have you experienced eyes, especially when it comes to the geoprocessing approaches and libraries, is for the other tasks: `raster_stats_task.py`, `road_density_task.py`, `site_buffering_task`. Each Task, as we discussed, essentially reads data, calls a set of functions in the `data_processing.py`module, and then saves the outputs to disk. Would be great to review both the tasks themselves, and the underlying functions, that hold the actual logic. 

I'd be curious to hear if you think that the current split between DAGs, Tasks and modules makes sense, in terms of being ETC and DRY (yes, I've started reading the book, it's good!).

The code itself is hopefully fairly well-structured, and the docstrings should generally be in good shape. I've made notes of some future improvements here and there. There are no unit tests yet, that's something that I'll have to add in the future. The book has made me question some old habits about commenting code, but that's a future discussion.

Finally, for the next PR I'll add some getting started steps to the README.

### CODE REVIEW

File: core/data/predicts.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+        gdf_site_coords = self.create_site_coord_geometries(df_concat)
+
+        # Save both dataframes to disk
+        df_concat.write_parquet(concat_data_output_path)
+        gdf_site_coords.to_file(site_coord_output_path)
+
+        runtime = str(timedelta(seconds=int(time.time() - start)))
+        logger.info(f"PREDICTS data processing finished in {runtime}.")
+
+    def concatenate_datasets(
```
Comment (dcdenu4) [inline core/data/predicts.py]: Does this function also exist in `data_processing`? Looks like some possible repeated code.

File: core/data/predicts.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+            raise ValueError(f"Extra columns in 'col_order': {extra_cols}")
+
+        # Reorder the columns according to 'col_order'
+        df_concat = df_concat.select(col_order)
+
+        logger.info("Finished concatenating datasets.")
+
+        return df_concat
+
+    def create_site_coord_geometries(self, df_concat: pl.DataFrame) -> gpd.GeoDataFrame:
```
Comment (dcdenu4) [inline core/data/predicts.py]: Looks like this function may have also been abstracted out to `data_processing`?

File: core/data/predicts_processing.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+    # Save both dataframes to disk
+    df_concat.write_parquet(configs.predicts.concat_data_output_path)
+    gdf_site_coords.to_file(configs.predicts.site_coord_output_path)
+
+    runtime = str(timedelta(seconds=(time.time() - start)))
+    logger.info(f"PREDICTS data processing finished in {runtime}.")
+
+
+def concatenate_datasets(
```
Comment (dcdenu4) [inline core/data/predicts_processing.py]: A 3rd implementation of this function? Sorry if I'm missing something!

Comment (dcdenu4) [reply core/data/predicts_processing.py]: Or now that I think about it, maybe you've already sorted these out in a different branch you were working on.

File: core/data/predicts_processing.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+from shapely.geometry import Point
+
+from core.utils.general_utils import create_custom_logger
+
+# Load config file into a box object and set up logging
+configs = Box.from_yaml(filename="data_configs.yaml")
+logger = create_custom_logger(__name__)
+
+
+def run() -> None:
```
Comment (dcdenu4) [inline core/data/predicts_processing.py]: Is this redundant with `predicts.run_task()`? Again, sorry if you've already sorted this all out.

File: core/data/predicts_task.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+# Load the config file into box object; ensure that it can be found regardless
+# of where the module is loaded / run from
+script_dir = os.path.dirname(os.path.abspath(__file__))
+config_path = os.path.join(script_dir, "data_configs.yaml")
+configs = Box.from_yaml(filename=config_path)
+
+logger = create_logger(__name__)
+
+
+class PredictsProcessingTask:
```
Comment (dcdenu4) [inline core/data/predicts_task.py:23]: Okay, I think this is the one that brings it all together and those other two modules are no longer needed!

**Review by dcdenu4 — COMMENTED** [review]
Sorry this review took so long, but everything looks really good! I had a few comments on what I think are just left over files / frameworks that you've probably already updated.

The only things that really caught my eye were calculating raster stats and handling multilinestrings for roads. You noted that some raster stats improvements can likely be made. For the road networks, I'd be curious if there was a way to handle those operations through a library without having to do it manually. But if it's working well it's also not something I'd spend any time looking into!

The code seems pretty DRY and orthogonal (changing one thing should have little impact on other parts). We've already chatted about how you're using the object oriented design and I liked seeing the general purpose functions lifted out into their own module.

File: core/data/predicts_task.py (hunk 5 above)
Comment (j-nystrom) [reply core/data/predicts_task.py:23]: Something weird going on. `predicts_task.py` is the one and only correct file. The others above don't exist anymore on my branch.

**Pull request merged**